modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/TransportSearchTemplateAction.java

```diff
@@ -110,6 +111,19 @@ static SearchRequest convert(SearchTemplateRequest searchTemplateRequest, Search
             builder.parseXContent(parser, false);
             builder.explain(searchTemplateRequest.isExplain());
             builder.profile(searchTemplateRequest.isProfile());
+            if (searchRequest.source() == null) {
+                searchRequest.source(new SearchSourceBuilder());
+            }
+            Integer trackTotalHitsUpTo = searchRequest.source().trackTotalHitsUpTo();
+            if (trackTotalHitsUpTo != null) {
+                if (builder.trackTotalHitsUpTo() == null) {
+                    builder.trackTotalHitsUpTo(trackTotalHitsUpTo);
+                } else if (builder.trackTotalHitsUpTo() != SearchContext.TRACK_TOTAL_HITS_ACCURATE
+                    && builder.trackTotalHitsUpTo() != SearchContext.TRACK_TOTAL_HITS_DISABLED) {
+                    throw new IllegalArgumentException("[" + RestSearchAction.TOTAL_HITS_AS_INT_PARAM + "] cannot be used " +
+                        "if the tracking of total hits is not accurate, got " + trackTotalHitsUpTo);
+                }
+            }
```

Comment:
When `rest_total_hits_as_int` is set to true, `trackTotalHitsUpTo`in the searchRequest is set to Integer.MAX_VALUE, 
we should copy the value to the new searchRequest here  and check whether the `track_total_hits` in the inline script is set to a not accurate number.